docs/source/rules.rst

```diff
@@ -33,14 +33,77 @@ Implementation
 .. autoclass:: sqlfluff.core.rules.base.LintFix
    :members:
 
+The `_eval` function of each rule takes a `RuleContext` parameter. It
+evaluate the given segment for violations, returning a `LintResult` if it
+finds an error. The `LintResult` includes a reference to the segment which
+"triggered" the error. Usually, it is the segment that needs correcting, **or**
+if the rule relates to something that is missing, then it should reference on
+the segment **following** the location where the missing element should be.
+
+Functional API
+--------------
+These newer modules provide a higher-level API for rules working with segments
+and slices. Rules that need to navigate or search the parse tree may benefit
+from using these. Eventually, the plan is for **all** rules to use these
+modules. At this time, the following rules use them:
+* `L001`
+* `L008`
+* `L009`
+* `L013`
+* `L015`
+* `L017`
+* `L021`
+* `L025`
+* `L031`
+* `L035`
+* `L036`
+* `L038`
+* `L042`
+* `L043`
+* `L047`
+* `L050`
+* `L052`
+
+.. autoclass:: sqlfluff.core.rules.functional.segments
+   :members:
+
+.. autoclass:: sqlfluff.core.rules.functional.segment_predicates
+   :members:
 
-The `_eval` function of each rule should take enough arguments that it can
-evaluate the position of the given segment in relation to its neighbors,
-and that the segment which finally "triggers" the error, should be the one
-that would be corrected OR if the rule relates to something that is missing,
-then it should flag on the segment FOLLOWING, the place that the desired
-element is missing.
+.. autoclass:: sqlfluff.core.rules.functional.raw_file_slices
+   :members:
+
+.. autoclass:: sqlfluff.core.rules.functional.raw_file_slice_predicates
+   :members:
```

Comment:
These autodoc commands don't seem to render properly at the moment: 
![image](https://user-images.githubusercontent.com/4670904/147781976-d2729d94-d317-465d-a4b2-673e9138f9a9.png)

Reply:
Fixes incoming...

Reply:
@barrywhart - can you push your most recent changes? I can't see them on the branch. I'll give them a spin when they're up.